File: crates/01-risk-sampler/c/bench_risk_sampler.c
/*
 * bench_risk_sampler.c
 *
 * Benchmark harness for the risk-sampler Monte Carlo simulation.
 *
 * Runs the simulation `ITERATIONS` times, measures wall-clock time with
 * clock_gettime(CLOCK_MONOTONIC), and emits a single JSON line:
 *
 *   {"ok": true, "mean_ns": <value>, "summary": "<text>", "iterations": 100}
 *
 * Compile with -O2 for a representative measurement.
 */

#define _DEFAULT_SOURCE
#include "risk_sampler.h"

#include <inttypes.h>
#include <stdio.h>
#include <stdint.h>
#include <time.h>

#define ITERATIONS 100
#define TRIALS     100000
#define SEED       42

int main(void)
{
    /* Build the same event portfolio used by the Rust main binary. */
    RiskEvent events[] = {
        { "Cyber attack",         0.05, 500000.0  },
        { "Server outage",        0.15,  50000.0  },
        { "Supply chain delay",   0.20,  25000.0  },
        { "Regulatory fine",      0.02, 1000000.0 },
    };
    uint64_t event_count = sizeof(events) / sizeof(events[0]);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (int i = 0; i < ITERATIONS; i++) {
        /* volatile to prevent the compiler from optimising away the call */
        volatile SimulationResult r = simulate(events, event_count, TRIALS, SEED);
        (void)r;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);

    int64_t elapsed_ns = (int64_t)(t1.tv_sec - t0.tv_sec) * 1000000000LL
                       + (int64_t)(t1.tv_nsec - t0.tv_nsec);
    int64_t mean_ns    = elapsed_ns / ITERATIONS;

    printf("{\"ok\": true, \"mean_ns\": %" PRId64 ", "
           "\"summary\": \"%d iterations of %d trials (seed %d) in %" PRId64 " ms, "
           "mean %" PRId64 " ns/iter\", \"iterations\": %d}\n",
           mean_ns,
           ITERATIONS, TRIALS, SEED,
           elapsed_ns / 1000000,
           mean_ns,
           ITERATIONS);

    return 0;
}

File: crates/01-risk-sampler/python/bench_risk_sampler.py
"""
Benchmark for risk_sampler.simulate -- prints a single JSON line.

Run with:  python bench_risk_sampler.py

The benchmark uses `timeit.repeat` which runs the target function multiple
times in a tight loop, then reports the *minimum* batch time. We convert
that to a mean per-call nanosecond figure so it's directly comparable to
Rust's criterion output.
"""

from __future__ import annotations

import json
import timeit

from risk_sampler import RiskEvent, simulate

# Match the portfolio from the Rust main.rs so the benchmark is comparable.
EVENTS: list[RiskEvent] = [
    RiskEvent(name="Cyber attack", probability=0.05, max_loss=500_000.0),
    RiskEvent(name="Server outage", probability=0.15, max_loss=50_000.0),
    RiskEvent(name="Supply chain delay", probability=0.20, max_loss=25_000.0),
    RiskEvent(name="Regulatory fine", probability=0.02, max_loss=1_000_000.0),
]

TRIALS: int = 100_000
SEED: int = 42
ITERATIONS: int = 100  # how many times timeit repeats the call


def _target() -> None:
    """Single invocation of simulate, used as the timeit target."""
    simulate(EVENTS, TRIALS, SEED)


def main() -> None:
    # `timeit.repeat` returns a list of total-seconds for each repeat batch.
    # With number=1 each entry is one call, and we take the best (minimum)
    # to reduce noise from OS scheduling jitter.
    times: list[float] = timeit.repeat(stmt=_target, number=1, repeat=ITERATIONS)

    best_s: float = min(times)
    mean_s: float = sum(times) / len(times)
    mean_ns: int = int(mean_s * 1_000_000_000)
    summary_ms: float = round(mean_s * 1_000, 1)

    result: dict = {
        "ok": True,
        "mean_ns": mean_ns,
        "summary": f"{TRIALS:,} trials in {summary_ms}ms",
        "iterations": ITERATIONS,
    }

    print(json.dumps(result))


if __name__ == "__main__":
    main()
